system/imc/loongson/loongson_la.c

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+//
+// Platform-specific code for Loongson LoongArch CPU
+//
+
+#include "boot.h"
+#include "cpuinfo.h"
+#include "memctrl.h"
+#include "memrw8.h"
+#include "memrw64.h"
+#include "cache.h"
+#include "cpuid.h"
+
+#include <larchintrin.h>
+
+#include "../imc.h"
+
+#define ENV_SIZE 0x100000 // 1MB
+
+#define MC_CONF_ADDRESS   0x800000000FF00000ULL
+#define CHIP_CONF_ADDRESS 0x800000001FE00000ULL
+
+bool route_flag;
+uint8_t max_mc;
+
+void read_imc_sequence(void)
+{
+    imc.tCL     = (uint16_t)read8((uint8_t *)(MC_CONF_ADDRESS + 0x1060));
+    imc.tCL_dec = 0;
+    imc.tRP     = (uint16_t)read8((uint8_t *)(MC_CONF_ADDRESS + 0x1006));
+    imc.tRCD    = (uint16_t)read8((uint8_t *)(MC_CONF_ADDRESS + 0x1047));
+    imc.tRAS    = (uint16_t)read8((uint8_t *)(MC_CONF_ADDRESS + 0x1040));
+
+    switch ((read8((uint8_t *)(MC_CONF_ADDRESS + 0x1024)) & 0x7)) {
+        case 0:
+              imc.width = 64;
+              break;
+        case 3:
+              imc.width = 16;
+              break;
+        case 5:
+              imc.width = 32;
+              break;
+        default:
+              imc.width = 0;
+              break;
+
+    }
+}
+
+bool read_imc_info(void)
+{
+    uint64_t fun_val;
+    uint8_t  i;
+    bool     ret = false;
+
+    if (route_flag) {
+        for (i = 0; i < max_mc; i++) {
+            fun_val  = read64((uint64_t *)(CHIP_CONF_ADDRESS | i << 16 | 0x180));
+            write64((uint64_t *)(CHIP_CONF_ADDRESS | i << 16 | 0x180), (fun_val & (~(1 << 4))));
+
+            if (read8((uint8_t *)(MC_CONF_ADDRESS)) == 0xFF || read8((uint8_t *)(MC_CONF_ADDRESS)) == 0x00) {
+                write64((uint64_t *)(CHIP_CONF_ADDRESS | i << 16 | 0x180), fun_val);
+                continue;
+            }
+
+            read_imc_sequence();
+
+            write64((uint64_t *)(CHIP_CONF_ADDRESS | i << 16 | 0x180), fun_val);
+            ret = true;
+            break;
+        }
+    } else {
+        for (i = 0; i < max_mc; i++) {
+            fun_val  = read64((uint64_t *)(CHIP_CONF_ADDRESS | 0x180));
+            write64((uint64_t *)(CHIP_CONF_ADDRESS | 0x180), (fun_val & (~(1 << (4 + (i * 5))))));
+
+            if (read8((uint8_t *)(MC_CONF_ADDRESS)) == 0xFF || read8((uint8_t *)(MC_CONF_ADDRESS)) == 0x00) {
+                write64((uint64_t *)(CHIP_CONF_ADDRESS | i << 16 | 0x180), fun_val);
+                continue;
+            }
+
+            read_imc_sequence();
+
+            write64((uint64_t *)(CHIP_CONF_ADDRESS | 0x180), fun_val);
+            ret = true;
+            break;
+        }
+    }
+
+    return ret;
+}
+
+void get_imc_config_loongson(void)
+{
+    uint32_t val;
+    uint16_t refc, loopc, div, div_mode, ref_clk;
+
+    imc.type  = "DDR4";
+
+    if (strstr(cpuid_info.brand_id.str, "3C") ||
+        strstr(cpuid_info.brand_id.str, "3D") ||
+        strstr(cpuid_info.brand_id.str, "3B6000")) {
```

Comment:
Consider `strncmp` to only match the prefix. Otherwise there's always the (unlikely but non-zero) chance that later Loongson processors are named according to a different scheme and a short fragment appears elsewhere, and the test will be a false-positive and cause problems.

Reply:
Let's me think about this.

Reply:
IMO, Loongson's CPU roadmap does not define other 3B6000 series except 3B6000 and 3B6000M. So next time I will add another logic to distinguish them.

Reply:
> IMO, Loongson's CPU roadmap does not define other 3B6000 series except 3B6000 and 3B6000M. So next time I will add another logic to distinguish them.

Isn't there 3B6600 on the roadmap?

Reply:
Yes, it is. But it probably be 2 MC not 4, and it will be an improved generation that will be presented with DDR5, this submit only for DDR4.